grpcbp/server.go

```diff
@@ -0,0 +1,148 @@
+package grpcbp
+
+import (
+	"net"
+	"time"
+
+	"github.com/reddit/baseplate.go"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/keepalive"
+)
+
+// ServerConfig is the argument struct for NewBaseplateServer. Please refer to
+// the documentation for each field to see how is it used.
+type ServerConfig struct {
+	// MaxConnectionIdle is a duration for the amount of time after which an idle
+	// connection would be closed by sending a GoAway. Idleness duration is
+	// defined since the most recent time the number of outstanding RPCs became
+	// zero or the connection establishment. The current default value is
+	// infinity.
+	MaxConnectionIdle time.Duration `yaml:"maxConnectionIdle"`
+
+	// MaxConnectionAge is a duration for the maximum amount of time a
+	// connection may exist before it will be closed by sending a GoAway. A
+	// random jitter of +/-10% will be added to MaxConnectionAge to spread out
+	// connection storms. The current default value is infinity.
+	MaxConnectionAge time.Duration `yaml:"maxConnectionAge"`
+
+	// MaxConnectionAgeGrace is an additive period after MaxConnectionAge after
+	// which the connection will be forcibly closed. The current defualt value is
+	// infinity.
+	MaxConnectionAgeGrace time.Duration `yaml:"maxConnectionAgeGrace"`
+
+	// After a duration of this time if the server doesn't see any activity it
+	// pings the client to see if the transport is still alive. If set below 1s,
+	// a minimum value of 1s will be used instead. The current default value is 2
+	// hours.
+	PingInactiveClientsAfter time.Duration `yaml:"pingInactiveClientsAfter"`
+
+	// After having pinged for keepalive check, the server waits for a duration
+	// of KeepAliveTimeout and if no activity is seen even after that the connection is
+	// closed. The current default value is 20 seconds.
+	KeepAliveTimeout time.Duration `yaml:"keepAliveTimeout"`
+
+	// KeepAliveMinTime is the minimum amount of time a client should wait before sending
+	// a keepalive ping.
+	KeepAliveMinTime time.Duration `yaml:"keepAliveMinTime"`
+
+	// If true, server allows keepalive pings even when there are no active
+	// streams(RPCs). If false, and client sends ping when there are no active
+	// streams, server will send GOAWAY and close the connection.
+	KeepAlivePermitWithoutStream bool `yaml:"keepAlivePermitWithoutStream"`
+}
+
+// Validate ensures all set values are within bound.
+func (s ServerConfig) Validate() error {
+	return nil
+}
+
+// GRPCServer provides passing in the generated gRPC service implementation and
+// reigster it on the created gRPC.Server.
+type GRPCServer interface {
```

Comment:
nit: since the name of this package is already `grpcbp`, `grpcbp.GRPCServer` feels redundant, we should rename this to just `Server` so it's `grpcbp.Server`.